impl/spanner/mutations/table.go

```diff
@@ -0,0 +1,311 @@
+	defer rtx.Close()
+
+	m, err := spanner.UpdateStruct(logTable, LogTableCols{
+		DirectoryID: directoryID,
+		LogID:       logID,
+		WriteToLog:  enabled,
+	})
+	if err != nil {
+		return err
+	}
+	_, err = t.client.Apply(ctx, []*spanner.Mutation{m})
+	return err
+}
+
+// AddLogs adds the logIDs to the list of active logs to send mutations to, and read mutations from.
+func (t *Table) AddLogs(ctx context.Context, directoryID string, logIDs ...int64) error {
+	if len(logIDs) == 0 {
+		return nil
+	}
+
+	rtx := t.client.Single()
+	defer rtx.Close()
+
+	mutations := make([]*spanner.Mutation, 0, len(logIDs))
+	for _, logID := range logIDs {
+		// Ignore already exists errors so this method can be retried.
+		m, err := spanner.InsertOrUpdateStruct(logTable, LogTableCols{
+			DirectoryID: directoryID,
+			LogID:       logID,
+			WriteToLog:  true,
+		})
+		if err != nil {
+			return err
+		}
+		mutations = append(mutations, m)
+	}
+	_, err := t.client.Apply(ctx, mutations)
+	return err
+}
+
+func (t *Table) ListLogs(ctx context.Context, directoryID string, writable bool) ([]int64, error) {
+	var stmt spanner.Statement
+	if writable {
+		stmt = spanner.NewStatement(
+			`SELECT LogID FROM LogStatus WHERE DirectoryID = @directory_id AND WriteToLog = TRUE`)
+	} else {
+		stmt = spanner.NewStatement(
+			`SELECT LogID FROM LogStatus WHERE DirectoryID = @directory_id`)
+	}
+	stmt.Params["directory_id"] = directoryID
+	var logIDs []int64
+	rtx := t.client.Single().WithTimestampBound(spanner.MaxStaleness(readStaleness))
+	defer rtx.Close()
+	if err := rtx.Query(ctx, stmt).Do(
+		func(row *spanner.Row) error {
+			var logID int64
+			if err := row.Columns(&logID); err != nil {
+				return err
+			}
+			logIDs = append(logIDs, logID)
+			return nil
+		}); err != nil {
+		return nil, err
+	}
+	if len(logIDs) == 0 {
+		return nil, status.Errorf(codes.NotFound, "not found")
+	}
+	return logIDs, nil
+}
+
+// Send submits an item to the queue and returns the commit timestamp.
+func (t *Table) Send(ctx context.Context, directoryID string, logID int64, entries ...*pb.EntryUpdate) (water.Mark, error) {
+	if len(entries) == 0 {
+		return water.Mark{}, nil
+	}
+	type Cols struct {
+		DirectoryID string
+		LogID       int64
+		Timestamp   time.Time
+		LocalID     int64
+		Mutation    []byte
+	}
+	ms := make([]*spanner.Mutation, 0, len(entries))
+	for i, e := range entries {
+		mBytes, err := proto.Marshal(e)
+		if err != nil {
+			return water.Mark{}, err
+		}
+		m, err := spanner.InsertStruct(mutTable, Cols{
+			DirectoryID: directoryID,
+			LogID:       logID,
+			Timestamp:   spanner.CommitTimestamp,
+			LocalID:     int64(i),
+			Mutation:    mBytes,
+		})
+		if err != nil {
+			return water.Mark{}, err
+		}
+		ms = append(ms, m)
+	}
+
+	commitTimestamp, err := t.client.ReadWriteTransaction(ctx,
+		func(_ context.Context, txn *spanner.ReadWriteTransaction) error {
+			return txn.BufferWrite(ms)
+		})
+	return timeToMark(commitTimestamp), err
```

Comment:
You may have investigated this and decided it's safe, but calling a function on something which may have been returned with an error seems like it might cause failures? Perhaps timeToMark works on any possible value of time.Time in which case discard.

Reply:
good call